File: src/client/panel/url-bar.js
import * as dom from "../dom";
import { Ref, empty } from "../../util/stream";


export const url = new Ref(null);

const top_style = dom.style({
  "left": "0px",
  "bottom": "0px",

  "white-space": "pre",
  // TODO maybe remove this
  "max-width": dom.calc("100%", "+", "1px"),

  "border-top-width": "1px",
  "border-right-width": "1px",
  "border-top-color": "black",
  "border-right-color": "black",
  "border-top-right-radius": "5px",

  //e.set("paddingTop", "0px")
  "padding-right": "2px", // 2px + 3px = 5px
  "padding-bottom": "1px",
  //e.set("padding-left", "2px")

  "color": "black",

  "background-color": "white",

  "box-shadow": "0px 0px 3px dimgray",
});

const text_style = dom.style({
  "margin-left": "3px",
  "margin-right": "3px"
});

const protocol_style = dom.style({
  "font-weight": "bold",
  "color": dom.hsl(120, 100, 25)
});

const domain_style = dom.style({
  "font-weight": "bold"
});

const path_style = dom.style({});

const file_style = dom.style({
  "font-weight": "bold",
  "color": "darkred" // TODO replace with hsl
});

const query_style = dom.style({
  "font-weight": "bold",
  "color": "darkred" // TODO replace with hsl
});

const hash_style = dom.style({
  "color": "darkblue" // TODO replace with hsl
});


const spacify = (x) =>
  x["replace"](/_|\-/g, " ");

// http://en.wikipedia.org/wiki/URI_scheme#Generic_syntax
const re_uri = /^([a-zA-Z][a-zA-Z0-9\+\.\-]*:)(?:(\/\/)([^\@]+\@)?([^\/\?\#\:]*)(\:[0-9]+)?)?([^\?\#]*?)([^\/\?\#]*)(\?[^\#]*)?(\#.*)?$/;

const parse = (x) => {
  const a = re_uri["exec"](x);
  if (a) {
    return {
      protocol:  a[1]["toLocaleLowerCase"](),
      separator: a[2] || null,
      authority: a[3] || null,
      domain:    a[4] || null,
      port:      (a[5]
                   ? +a[5]
                   : null),
      path:      a[6] || null,
      file:      a[7] || null,
      query:     a[8] || null,
      hash:      a[9] || null
    }
  } else {
    throw new Error("Invalid URI: " + x)
  }
};

const simplify = (x) => {
  const y = {};

  if (x.protocol === "http:" || x.protocol === "https:") {
    y.protocol  = null;
    y.separator = null;

  } else {
    y.protocol  = x.protocol;
    y.separator = x.separator;
  }

  y.authority = x.authority;

  if (x.domain === null) {
    y.domain = x.domain;

  } else {
    // http://en.wikipedia.org/wiki/List_of_Internet_top-level_domains
    y.domain = x.domain["replace"](/^www\.|\.\w\w$/g, "") // .co.uk
                       ["replace"](/\.(?:aero|asia|biz|cat|com|co|coop|info|int|jobs|mobi|museum|name|net|org|pro|tel|travel|xxx|edu|gov|mil)$/, "")
                       // TODO: is this needed?
                       ["replace"](/\.\w\w$/, "") // .ac.edu
  }

  y.port  = x.port;
  y.path  = x.path;
  y.file  = x.file;
  y.query = x.query;
  y.hash  = x.hash;

  return y;
};

const minify = (x) => {
  const y = simplify(parse(x));

  const query = y.query;
  const path  = y.path;
  const file  = y.file;
  //const hash  = y.hash;

  y.query = null;
  y.path  = null;
  y.file  = null;
  //y.hash  = null;

  if (query) {
    y.query = query["replace"](/^\?/, "");
    y.query = y.query["replace"](/^[\+&;]/, "");
    y.query = y.query["replace"](/[\+&;]/g, ", ");
    y.query = y.query["replace"](/=/g, ":");
    y.query = spacify(decodeURIComponent(y.query));

  } else if (file) {
    y.file = spacify(decodeURIComponent(file["replace"](/\.(?:html?|php|asp)$/, "")));

  } else if (path) {
    y.path = spacify(decodeURIComponent(path));
    if (y.path === "/") {
      y.path = null;
    }
  }

  /*if (hash) {
    y.hash = spacify(decodeURIComponent(hash["replace"](/^\#/, "")));
  }*/

  return y;
};

const parsed_url = url.keep((x) => x !== null).map(({ url }) => minify(url));

const make = (style, f) =>
  dom.row((e) => {
    e.add_style(text_style);
    e.add_style(style);

    const text = dom.text("");

    e.push(text);

    return parsed_url.map((x) => {
      const s = f(x);

      if (s === null) {
        e.hide();

      } else {
        text.text = s;
        e.show();
      }
    });
  });

dom.floating((e) => {
  e.add_style(top_style);

  // TODO check if any of these need "flex-shrink": 1
  e.push(dom.row((e) => {
    e.push(make(protocol_style, (x) => x.protocol));
    e.push(make(domain_style, (x) => x.domain));
    e.push(make(path_style, (x) => x.path));
    e.push(make(file_style, (x) => x.file));
    e.push(make(query_style, (x) => x.query));
    e.push(make(hash_style, (x) => x.hash));
    return empty;
  }));

  return url.map((o) => {
    if (o === null) {
      e.hide();

    } else {
      e.show();

      const box = e.get_position();

      if (o.x <= box.right && o.y >= box.top) {
        e.hide();
      }
    }
  });
});

File: src/bin/sidebar/url_bar.rs
use tab_organizer::{RegExp, decode_uri_component};

lazy_static! {
    // http://en.wikipedia.org/wiki/URI_scheme#Generic_syntax
    // TODO test this
    static ref URL_REGEXP: RegExp = RegExp::new(
        "^([a-zA-Z][a-zA-Z0-9\\+\\.\\-]*:)(?:(\\/\\/)([^\\@]+\\@)?([^\\/\\?\\#\\:]*)(\\:[0-9]+)?)?([^\\?\\#]*?)([^\\/\\?\\#]*)(\\?[^\\#]*)?(\\#.*)?$",
        ""
    );

    // http://en.wikipedia.org/wiki/List_of_Internet_top-level_domains
    // TODO test this
    static ref DOMAIN_REGEXP: RegExp = RegExp::new(
        "(?:^www?\\.)|(?:(?:\\.[a-z][a-z])?\\.[a-z]+$)",
        "gi"
    );

    static ref SPACIFY_REGEXP: RegExp = RegExp::new(
        "[_\\-\\n\\r]",
        "g"
    );

    static ref FILE_REGEXP: RegExp = RegExp::new(
        "\\.(?:html?|php|asp)$",
        "g"
    );

    static ref QUERY_REMOVE_REGEXP: RegExp = RegExp::new(
        "^\\??[\\+&;]?",
        ""
    );

    static ref QUERY_SEPARATOR_REGEXP: RegExp = RegExp::new(
        "[\\+&;]",
        "g"
    );

    static ref QUERY_KEY_VALUE_REGEXP: RegExp = RegExp::new(
        "=",
        "g"
    );
}

pub(crate) struct UrlBar {
    pub(crate) protocol: Option<String>,
    pub(crate) separator: Option<String>,
    pub(crate) authority: Option<String>,
    pub(crate) domain: Option<String>,
    pub(crate) port: Option<String>,
    pub(crate) path: Option<String>,
    pub(crate) file: Option<String>,
    pub(crate) query: Option<String>,
    pub(crate) hash: Option<String>,
}

fn minify_domain(domain: String) -> String {
    DOMAIN_REGEXP.replace(&domain, "")
}

fn spacify(input: &str) -> String {
    SPACIFY_REGEXP.replace(input, " ")
}

impl UrlBar {
    pub(crate) fn new(url: &str) -> Option<Self> {
        let parsed = URL_REGEXP.first_match(url);

        parsed.map(|mut parsed| {
            // TODO is there a better way of doing this ?
            let protocol = parsed.remove(1); // TODO lower case this ?
            let separator = parsed.remove(1);
            let authority = parsed.remove(1);
            let domain = parsed.remove(1);
            let port = parsed.remove(1);
            let path = parsed.remove(1);
            let file = parsed.remove(1);
            let query = parsed.remove(1);
            let hash = parsed.remove(1);

            Self { protocol, separator, authority, domain, port, path, file, query, hash }
        })
    }

    pub(crate) fn minify(self) -> Self {
        let should_protocol = match self.protocol.as_ref().map(|x| x.as_str()) {
            Some("https:") | Some("http:") | Some("") => false,
            _ => true,
        };

        let old_query = match self.query.as_ref().map(|x| x.as_str()) {
            Some("?") | Some("") => None,
            _ => self.query,
        };

        let mut path = None;
        let mut file = None;
        let mut query = None;

        if let Some(old_query) = old_query {
            query = Some({
                let query = QUERY_REMOVE_REGEXP.replace(&old_query, "");
                let query = QUERY_SEPARATOR_REGEXP.replace(&query, ", ");
                let query = QUERY_KEY_VALUE_REGEXP.replace(&query, ":");
                let query = spacify(&decode_uri_component(&query));
                query
            });

        } else if let Some(old_file) = self.file {
            file = Some(spacify(&decode_uri_component(&FILE_REGEXP.replace(&old_file, ""))));

        } else if let Some(old_path) = self.path {
            if old_path != "/" {
                path = Some(spacify(&decode_uri_component(&old_path)));
            }
        }

        let hash = match self.hash.as_ref().map(|x| x.as_str()) {
            Some("#") | Some("") => None,
            _ => self.hash.map(|hash| spacify(&decode_uri_component(&hash))),
        };

        Self {
            protocol: if should_protocol { self.protocol } else { None },
            separator: if should_protocol { self.separator } else { None },
            authority: self.authority,
            domain: self.domain.map(minify_domain),
            port: self.port,
            path,
            file,
            query,
            hash,
        }
    }
}
